oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,360 @@
+   * modification causes many intermediate modification states), making results more
+   * "accurate"/comparable as well as allowing faster processing of geometries.
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param osmEntityFilter only timestamps for which the entity matches this filter are returned
+   * @param changesetTimestamps association between timestamps and changeset-ids, can be obtained
+   *        from oshEntity by calling {@link #getChangesetTimestamps}.
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh,
+      Predicate<OSMEntity> osmEntityFilter, Map<OSHDBTimestamp, Long> changesetTimestamps) {
+    List<OSHDBTimestamp> allModificationTimestamps = getModificationTimestamps(osh, true, osmEntityFilter);
+
+    if (allModificationTimestamps.size() <= 1) {
+      return allModificationTimestamps;
+    }
+    // group modification timestamps by changeset
+    List<OSHDBTimestamp> result = new ArrayList<>(allModificationTimestamps.size());
+    allModificationTimestamps = Lists.reverse(allModificationTimestamps);
+    long nextChangeset = -1L;
+    for (OSHDBTimestamp timestamp : allModificationTimestamps) {
+      long changeset = changesetTimestamps.get(timestamp);
+      if (changeset != nextChangeset) {
+        result.add(timestamp);
+      }
+      nextChangeset = changeset;
+    }
+
+    return Lists.reverse(result);
+  }
+
+  /**
+   * Returns all timestamps at which this entity (or one or more of its child entities) has been
+   * modified and matches a given condition/filter.
+   * 
+   * @param osh the osh entity to work on
+   * @param osmEntityFilter only timestamps for which the entity matches this filter are returned
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh,
+      Predicate<OSMEntity> osmEntityFilter) {
+    return getModificationTimestamps(osh, true, osmEntityFilter);
+  }
+
+  private static List<OSHDBTimestamp> getModificationTimestamps(
+      OSHEntity osh, boolean recurse, Predicate<OSMEntity> osmEntityFilter) {
+    // first, store this ways direct modifications (i.e. the major versions' timestamps)
+    var entityTimestamps = getModificationTimestampsReverseNonRecursed(osh, osmEntityFilter);
+    if (!recurse || osh instanceof OSHNode) {
+      return Lists.reverse(entityTimestamps);
+    }
+    // recurse members: start by collecting all referenced members' validity time intervals
+    var membersValidityTimes = collectMembershipTimeIntervals(osh, osmEntityFilter);
+    // fill in member modification timestamps
+    var result = fillMembersModificationTimestamps(membersValidityTimes);
+    // combine with the entity's own modification timestamps
+    result.addAll(entityTimestamps);
+    return new ArrayList<>(result);
```

Comment:
what caught my eye here, is that in the non-recurse or node case we return a reversed list. But not in the other case. I wonder if this is correct.

Reply:
That _should_ not be the case, since:

* in the `!recurse || osh instanceof OSHNode` case the result is `Lists.reverse` of the result from `getModificationTimestampsReverseNonRecursed` which is in reverse order :arrow_right: we return it in the "natural" order.
* otherwise the `result` is a list generated from a `SortedSet` (result) which takes care of sorting for us. :arrow_right: same "natural" order as above

Maybe it would be better to not use `var result = …`  here, so it's clearer that the result stems from a sorted set? What do think?